### PR TITLE
Backport recent toolchain related fixes

### LINKED ISSUE #23025: Back-port toolchain related fixes to 7.6.1
Need to backport following changes:

- #22819 
- #23024 
- #23050 
- #23112

### PR DESCRIPTION
**Pull request by jbartok**
Fixes #23025

### CODE REVIEW

**Comment by jbartok** [comment]
@bot-gradle test QFL

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationInternal.java [hunk 1]
```diff
@@ -36,7 +37,7 @@
 import java.util.Set;
 import java.util.function.Supplier;
 
-public interface ConfigurationInternal extends ResolveContext, Configuration, DeprecatableConfiguration, DependencyMetaDataProvider {
+public interface ConfigurationInternal extends ResolveContext, Configuration, DeprecatableConfiguration, DependencyMetaDataProvider, FinalizableValue {
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/ConfigurationInternal.java:40]: Given this is a backport, I would exclude this change, just to be on the safe side.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java [hunk 2]
```diff
@@ -46,7 +47,7 @@
 import java.util.List;
 import java.util.Set;
 
-public class DefaultConfigurationPublications implements ConfigurationPublications {
+public class DefaultConfigurationPublications implements ConfigurationPublications, FinalizableValue {
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationPublications.java:50]: I would exclude that change as well.

File: subprojects/core/src/main/java/org/gradle/api/internal/SettingsInternal.java [hunk 3]
```diff
@@ -25,12 +25,13 @@
...
 import org.gradle.initialization.DefaultProjectDescriptor;
 import org.gradle.initialization.IncludedBuildSpec;
+import org.gradle.internal.FinalizableValue;
 import org.gradle.internal.management.DependencyResolutionManagementInternal;
 import org.gradle.internal.service.ServiceRegistry;
 
 import java.util.List;
 
-public interface SettingsInternal extends Settings, PluginAwareInternal {
+public interface SettingsInternal extends Settings, PluginAwareInternal, FinalizableValue {
```
Comment (ljacomet) [inline subprojects/core/src/main/java/org/gradle/api/internal/SettingsInternal.java:34]: That one as well

**Review by ljacomet — CHANGES_REQUESTED** [review]
Looks good overall, and nice that it does not contain all the changes for abstract methods in DefaultSettings.

However, I would also not backport other changes, see below:

**Review by ljacomet — APPROVED** [review]

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Pull request merged**